proving_system/stark/src/air/example/cairo.rs

```diff
@@ -0,0 +1,305 @@
+        };
+
+        let last_step = num_steps - 1;
+
+        let pub_inputs = PublicInputs {
+            pc_init: trace.get(0, FRAME_PC),
+            ap_init: trace.get(0, FRAME_AP),
+            fp_init: trace.get(0, FRAME_FP),
+            pc_final: trace.get(last_step, FRAME_PC),
+            ap_final: trace.get(last_step, FRAME_AP),
+            num_steps,
+        };
+
+        Self {
+            context,
+            pub_inputs,
+        }
+    }
+}
+
+impl AIR for CairoAIR {
+    type Field = Stark252PrimeField;
+
+    fn compute_transition(&self, frame: &Frame<Self::Field>) -> Vec<FieldElement<Self::Field>> {
+        let mut constraints: Vec<FieldElement<Self::Field>> =
+            vec![FE::zero(); self.num_transition_constraints()];
+
+        compute_instr_constraints(&mut constraints, frame);
+        compute_operand_constraints(&mut constraints, frame);
+        compute_register_constraints(&mut constraints, frame);
+        compute_opcode_constraints(&mut constraints, frame);
+        enforce_selector(&mut constraints, frame);
+
+        constraints
+    }
+
+    /// From the Cairo whitepaper, section 9.10.
+    /// These are part of the register constraints.
+    ///
+    /// Boundary constraints:
+    ///  * ap_0 = fp_0 = ap_i
+    ///  * ap_t = ap_f
+    ///  * pc_0 = pc_i
+    ///  * pc_t = pc_f
+    fn boundary_constraints(&self) -> BoundaryConstraints<Self::Field> {
+        let last_step = self.context.trace_length - 1;
+
+        let initial_pc =
+            BoundaryConstraint::new(MEM_A_TRACE_OFFSET, 0, self.pub_inputs.pc_init.clone());
+        let initial_ap =
+            BoundaryConstraint::new(MEM_P_TRACE_OFFSET, 0, self.pub_inputs.ap_init.clone());
+
+        let final_pc = BoundaryConstraint::new(
+            MEM_A_TRACE_OFFSET,
+            last_step,
+            self.pub_inputs.pc_final.clone(),
+        );
+        let final_ap = BoundaryConstraint::new(
+            MEM_P_TRACE_OFFSET,
+            last_step,
+            self.pub_inputs.ap_final.clone(),
+        );
+
+        let constraints = vec![initial_pc, initial_ap, final_pc, final_ap];
+
+        BoundaryConstraints::from_constraints(constraints)
+    }
+
+    fn context(&self) -> AirContext {
+        self.context.clone()
+    }
+}
+
+/// From the Cairo whitepaper, section 9.10
+fn compute_instr_constraints(constraints: &mut [FE], frame: &Frame<Stark252PrimeField>) {
+    // These constraints are only applied over elements of the same row.
+    let curr = frame.get_row(0);
+
+    // Bit constraints
+    for (i, flag) in curr[0..16].iter().enumerate() {
+        constraints[i] = match i {
+            0..=14 => flag * (flag - FE::one()),
+            15 => flag.clone(),
+            _ => panic!("Unknown flag offset"),
+        };
+    }
+
+    // Instruction unpacking
+    let two = FE::from(2);
+    let b16 = two.pow(16u32);
+    let b32 = two.pow(32u32);
+    let b48 = two.pow(48u32);
+
+    let f0 = &curr[0..15]
```

Comment:
Can we call this `f0_tilde` or `f0_squiggle` to match the notation of the paper?

Reply:
good suggestion, done!